File: js/monadic-laws.js
var assert = require('assert');

var laws = function() {
  var firstMonadicLaw = function(bind, wrap, f, value) {
    assert.equal(bind(wrap(value), f)(),
                 f(value)());
  };
  
  var secondMonadicLaw = function(bind, wrap, value) {
    assert.equal(bind(wrap(value), wrap)(),
                 wrap(value)());
  };

  var thirdMonadicLaw = function(bind, wrap, f, g, value) {
    var mv = wrap("candy");

    assert.equal(bind(bind(mv,f),g)(),
                 bind(mv,function(x) { return bind(f(x),g) })());
  }

  return {
    first: firstMonadicLaw,
    second: secondMonadicLaw,
    third: thirdMonadicLaw
  }
}();

exports.laws = laws;

File: clojure/test/monads_blog/test/monadic_laws.clj
(ns monads-blog.test.monadic-laws
  (:use [clojure.test]))

(defn first-monadic-law [wrap bind f value]
  (= ((bind (wrap value) f)) ((f value))))

(defn second-monadic-law [wrap bind value]
  (= ((bind (wrap value) wrap)) ((wrap value))))

(defn third-monadic-law [wrap bind f g value]
  (let [mv (wrap value)]
    (= ((bind (bind mv f) g)) ((bind mv (fn[x] (bind (f x) g)))))))
